### DISCUSSION TITLE
What is stopping us from using composer autoloader all the way?

### DISCUSSION THREAD

**Discussion by func0der** [OpenMage/magento-lts#1871] [Ideas]
Hey there,

I was wondering what is stopping us from using the Composer Autoloader for everything in OpenMage and get rid of the `Varien_Autoload` completely.

This should give us a boost in performance on production, right? Since on prod we would have a class map and no potentially failing `includes`.
And since all modules are placed in `app/code` and/or `lib` (also the paths that are set as include paths in `app/Mage.php`) on installation, which is done by `composer` most of the time, we would be able to use classmap locally, too.

Only downside would be, that local development would get a little bit more annoying, since you'd have to run `composer dump-autoload` all the time after creating a new file.

Since the composer autoloader comes first when trying to find a class (at least in the current setup), we could use the classmap AND the `Varien_Autoload`. This should only stay for a grace period though, since it gets pretty confusing fast.

What do you guys think?

---
Edit: Just found another issue. We have the possible overwrites of `community` and `core` classes in `local`, which could potentially confuse `composer`.
This can be solved by defining the folders in `classmap` in the right order:

```json
  "autoload": {
    "classmap": [
      "webroot/app/code/local",
      "webroot/app/code/community",
      "webroot/app/code/core",
      "webroot/lib"
    ]
  },
```

**Comment by Flyingmana** [comment]
There might be issues (unexpected behaviors) with classes not following the Magento Namespace pattern.
Also we have some people, who use openmage without composer, and we have not finished yet providing a full build download which already contains the vendor.
But in total, the performance gain is probably less then 10ms.

Although, there is nothing which prevents you from using the composer classmap currently and gaining some statistics and possible issues which only become clear in real usage. So if we can help with anything related to testing this, keep us posted.

**Comment by func0der** [comment]
Okay. Thank you for the insight.

> There might be issues (unexpected behaviors) with classes not following the Magento Namespace pattern.

Could you elaborate on this, please?
If classes do not follow the Magento Namespace pattern, they would not be loaded either way, would they?

> Also we have some people, who use openmage without composer...

I did not think about people using the software as a download. That could be a problem.

> ...and we have not finished yet providing a full build download which already contains the vendor.

How would such a package look like? Just a `vendor` folder containing the autoloader files and maybe some libraries that were removed from `lib` and instead required via `composer`?

> But in total, the performance gain is probably less then 10ms.

I only got around to mention this issue, because there are some modules out there that try to implement autoloading themselves, because they deploy Code in `lib` that is basically taken from Packagist/GitHub and extracted there. All of that code is PSR-4 based or at least PSR-0.
Sometimes these vendors did an okay job of implementing an autoloader that is prefix based and only triggers if 'their' classes are requested: https://github.com/OpenMageModuleFostering/dhl_versenden/blob/master/app/code/community/Dhl/Versenden/Helper/Autoloader.php
Other did a not so great job and query the filesystem for EVERY class that is requested during a request: https://github.com/OpenMageModuleFostering/iways_paypalplus/blob/master/app/code/community/Iways/PayPalPlus/Model/Autoloader.php#L58
And since those autoloaders are registered during the OpenMage lifecycle (https://github.com/OpenMageModuleFostering/iways_paypalplus/blob/master/app/code/community/Iways/PayPalPlus/etc/config.xml#L34-L43) they are at the top of the stack.

I know that this is not an issue of `OpenMage` as is, but things like that probably happen, because Magento has drifted so far away from modern standards back in its time.
The PayPalPlus autoloader alone has costs me ~400ms per request.

> Although, there is nothing which prevents you from using the composer classmap currently and gaining some statistics and possible issues which only become clear in real usage. 

Realized that while typing, too :D 
I will try it. I am trying to optimize a website right now and though I am having a lot of issues to figure out right now, I try to keep documentation in mind :D
